### PR TITLE
Solution added

### PR DESCRIPTION
**Pull request by kmety**

### CODE REVIEW

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 1]
```diff
@@ -3,48 +3,155 @@
...
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
-    @Override
-    public void add(T value) {
+    private int size = 0;
+    private Node<T> first;
+    private Node<T> last;
+
+    private static class Node<T> {
+        T item;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: add access modifier

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 2]
```diff
@@ -3,48 +3,155 @@
...
+        Node<T> next;
+        Node<T> prev;
 
+        Node(Node<T> prev, T element, Node<T> next) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: the same

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 3]
```diff
@@ -3,48 +3,155 @@
...
+            this.next = next;
+            this.prev = prev;
+        }
     }
 
-    @Override
-    public void add(T value, int index) {
+    public void add(T value) {
+        linkLast(value);
+    }
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: add override

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 4]
```diff
@@ -3,48 +3,155 @@
...
+                    return unlink(x);
+                }
+            }
+        }
         return null;
     }
 
-    @Override
+    private void linkLast(T element) {
+        final Node<T> l = last;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: CS

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 5]
```diff
@@ -3,48 +3,155 @@
...
+        final Node<T> newNode = new Node<>(l, element, null);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: why final?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 6]
```diff
@@ -3,48 +3,155 @@
...
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(node(index));
     }
 
-    @Override
     public T remove(T t) {
+        if (t == null) {
+            for (Node<T> x = first; x != null; x = x.next) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: can we simplify this method and use one loop?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 7]
```diff
@@ -3,48 +3,155 @@
...
+        currentNode.prev = newNode;
+        if (pred == null) {
+            first = newNode;
+        } else {
+            pred.next = newNode;
+        }
+        size++;
+    }
+
+    private T unlink(Node<T> node) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: better rename to removeElement()

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 8]
```diff
@@ -3,48 +3,155 @@
...
+            prev.next = next;
+            node.prev = null;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+        node.item = null;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: we can write:
node = null
and then remove node.prev = null; and node.next = null;

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 9]
```diff
@@ -3,48 +3,155 @@
...
+        size--;
+        return element;
+    }
+
+    private Node<T> node(int index) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
    private Node<T> getNode(int index) {
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 10]
```diff
@@ -3,48 +3,155 @@
...
+        if (index < (size >> 1)) {
+            Node<T> x = first;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: CS

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 11]
```diff
@@ -8,97 +8,90 @@
...
-            }
-        } else {
-            for (Node<T> x = first; x != null; x = x.next) {
-                if (t.equals(x.item)) {
-                    return unlink(x);
-                }
+    @Override
+    public T remove(T value) {
+        for (Node<T> node = first; node != null; node = node.next) {
+            if ((value == null && node.item == null) || node.item.equals(value)) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: (value == node.item) || node.item.equals(value)

Comment (kmety) [reply src/main/java/core/basesyntax/MyLinkedList.java]: Got it)

**Pull request closed without merging**